### PR TITLE
CXF-8557: Incorrect Proxy Path Segmenting when @Path Annotation RegexExpression Contains '/'

### PR DESCRIPTION
**Pull request by reta**
A service in question needs to be reachable via multiple paths. The @Path parameter allows for regex matching on a class-level like the following: @Path("/{a: regex}"). A class-level `@PathParam` is created by the name of a. On the server side, the following annotations both correctly receive requests at `/foo/bar`:

```
@Path("/{a : foo/bar}")
@Path("/{a : foo\\/bar}")
```

However, on the client side, the following URI(s)
```
final URI uri = UriBuilder
            .fromUri("/")
            .path("{a : foo/bar}")
            .buildFromEncoded("foo/bar");
```
are not properly constructed due to the fact that `/` is used as path segment separator without taking into account the URI template expression.

### CODE REVIEW

**Comment by reta** [comment]
@andymc12 if you have a moment please to take a look, much appreciate it

File: rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/utils/JAXRSUtils.java [hunk 1]
```diff
@@ -181,20 +181,51 @@ public static boolean isStreamingOutType(Class<?> type) {
...
+                    final String segment = thePath.substring(start, i);
+                    segments.add(new PathSegmentImpl(segment, decode));
+                }
+                
+                // advance the positions, empty path segments
+                start = i + 1;
+            } else if (thePath.charAt(i) == '{') {
+                ++depth;
+            } else if (thePath.charAt(i) == '}') {
+                --depth; // could go negative, since the template could be unbalanced
```
Comment (andymc12) [inline rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/utils/JAXRSUtils.java:215]: if it goes negative, should we throw an IllegalArgumentException?

**Review by andymc12 — APPROVED** [review]
@reta - I think this looks good.  The one thing that I'm sure about is what the expected behavior is when the unclosed curly braces are used. I lean toward throwing an `IllegalArgumentException` unless they are escaped, but I'm not really sure that that is the correct behavior either, so handling them as part of the path seems reasonable too.

Thanks for finding and fixing this!

File: rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/utils/JAXRSUtils.java (hunk 1 above)
Comment (reta) [reply rt/frontend/jaxrs/src/main/java/org/apache/cxf/jaxrs/utils/JAXRSUtils.java:215]: Thanks a lot for looking, @andymc12 . I was also not sure what the correct behavior should be. I think throwing the exception in this case would be surprising to the users (we may break existing code), however I think we have to address this issue in any case. I would suggest to fallback to straightforward `/` separator (current implementation) if we detect in unparsable template definitions (like unbalanced curly braces). What do you think?

**Pull request merged**